statemachine/statemachine.py

```diff
@@ -159,9 +162,35 @@ def allowed_events(self):
             for event in self.current_state.transitions.unique_events
         ]
 
-    def _process(self, trigger):
-        """This method will also handle execution queue"""
-        return trigger()
+    def _process(self, trigger):  # noqa: C901
+        # Check if the machine is running in "queued" mode
+        if not self._queued:
+            # If not, check if the queue is empty
+            if not self._external_queue:
+                # If the queue is empty, execute the trigger immediately
+                return trigger()
+            # If the queue is not empty, raise an exception as something is wrong
+            raise TransitionNotAllowed("Queue not empty.")
+
+        # If the machine is in "queued" mode, add the trigger to the queue
+        self._external_queue.append(trigger)
+        # If there is already a trigger in the queue,
+        # only return as the processing loop should already be running
+        if len(self._external_queue) > 1:
+            return
```

Comment:
I don't get what is the use case for that. How the client will know that there is already something running?

Reply:
A bold statement is that: On queued mode, you don't get results back when triggering an event.

So think on this return as a mutex, only the first event on the queue will enter the processing loop. If while this event is processed other events are put on the queue, they don't will be processed immediately (thus the return) and will be evaluated on the loop in the first processing loop.

Maybe we should implement this as an explicit mutex: Only the first event will reach the critical section of the loop processing.

What do you think?